Add Media File and Attachment Page options to the native Image block Link To menu

dcalhoun opened the pull request:
## Related PRs

* https://github.com/wordpress-mobile/gutenberg-mobile/pull/3971
* [No review needed] https://github.com/wordpress-mobile/WordPress-iOS/pull/17174
* [No review needed]  https://github.com/wordpress-mobile/WordPress-Android/pull/15336

## Description
Fixes https://github.com/wordpress-mobile/gutenberg-mobile/issues/924. Add Media File and Attachment Page options to the native Image block Link To menu to allow quickly setting commonly used URLs. This feature already exists in the web interface. 

Installable builds are available on the aforementioned PRs. 

## How has this been tested?

### Verify Link Destinations
1. Add an Image block.
2. Set the media to an existing upload or upload a new image.
3. Tap "Link to."
4. Select one of the options.
5. ℹ️ **Expected:** The additional "Open in new tab" and "Link Rel" options are displayed.
5. Publish and visit a preview of the post.
6. ℹ️ **Expected:** The image is linked to the selected option. 
7. Repeat steps 4-7 with each of the other options. 

### Verify External Image Options
1. Add an Image block.
2. Set the media to an external URL.
3. Tap "Link to."
4. ℹ️ **Expected:** The "Attachment Page" option is not available. 

## Screenshots <!-- if applicable -->

<details><summary>Screenshots</summary>



| Android | iOS |
| - | - |
| ![android-link-to](https://user-images.githubusercontent.com/438664/134232099-2d625f73-8ccd-495a-af4f-bd8ed1999df5.jpg) | ![ios-link-to](https://user-images.githubusercontent.com/438664/134233531-4be44a94-345d-4f7d-a368-e202141d5cc7.jpeg) |
| ![android-none](https://user-images.githubusercontent.com/438664/134232141-ef8380f2-90d7-4131-9299-44ad1b42226c.jpg) | ![ios-none](https://user-images.githubusercontent.com/438664/134233560-2170d367-8b49-4de8-9343-e239ec2051a4.PNG) |
| ![android-custom-url](https://user-images.githubusercontent.com/438664/134232180-4c4636d9-9b2e-4f56-9c49-28f4c82357c8.jpg) | ![ios-custom-url](https://user-images.githubusercontent.com/438664/134233591-51f96fbf-4152-4d05-aebb-ee29c8353775.PNG) |

</details>

<details><summary>Video</summary>

https://user-images.githubusercontent.com/438664/134233479-8d124a1f-65f2-4d63-b549-24dd048c2cf3.MP4

</details>

## Types of changes
New feature

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

dcalhoun committed:
Replace style-based key with index key

Stringifying the styles appears unnecessary, and it was also causing
missing key warnings in tests as the styles are undefined in that
context. Leveraging the index appears to be enough to satisfy both the
app and the tests.

dcalhoun committed:
Refactor existing Image edit tests to avoid referencing internals

Test from the "public" API of the component and its output, rather than
internal methods. I.e. press rendered buttons instead of calling methods
by name.

dcalhoun committed:
Remove unnecessary native file extension reference

Metro resolves native files by default.

dcalhoun committed:
Avoid global act by awaiting specific change within Image component

The Image component invokes `getMedia` within its selector. This results
in an async resolver running, which leads to changes to the component
after the async work completes. Rather than wrapping the entire test
helper with `act`, we should target the specific change within the test
file.

dcalhoun committed:
Add missing React Navigation testing setup

Per React Navigation documentation, this file should be included.
https://bit.ly/3n6Rotv

dcalhoun committed:
Add note regarding React Navigation async test error

There appears to be a bug in React Navigation that causes error logs in
Jest tests. We must await the `fireEvent` that triggers the navigation
event in order to suppress the error. https://git.io/Ju35Z

dcalhoun committed:
Add initial UI for Image Link To options

Displays "Media" and "Attachment Page" options within the Link To
settings for the Image block. The logic does not work for the new
options, and the implementation is rudimentary. Several `TODO(David)`
are in place for remaining work.

dcalhoun committed:
Intercept link settings navigation with callback instead of string

Using a callback avoids placing context-specific props reference inside
of link settings. I.e. `imageUrl` and `attachmentPageUrl` are no longer
referenced within link settings.

dcalhoun committed:
Align labels with web UI

Reusing label text from the web will likely increase familiarity and
reduce confusion for users.

dcalhoun committed:
Display selected link to option in Image block settings

Display the currently selected link to option in the top-level Image
block settings.

dcalhoun committed:
Reset link destination to None when removing Custom URL

dcalhoun committed:
Indicate selected link to option with check mark icon

dcalhoun committed:
Fix comment typo

dcalhoun commented:
@mkevins this is very much still a work-in-progress and not completely functional, but I wanted to share it in case you'd like to provide a high-level review on the overall approach. A review now is not urgent if you do not have time, I will definitely seek your full review before merging this work. Thanks!

dcalhoun committed:
Display correct value mask when URL value is set

Relying upon the `placeholder` did not work correctly whenever a `value`
was set, i.e. in the case of a Custom URL. Additionally, overloading the
`placeholder` prop is confusing. The new `valueMask` prop hopefully
better communicates the intent.

dcalhoun committed:
Reference Cell attached to BottomSheet component

Avoid superfluous import.

dcalhoun committed:
Fix incorrect prop references

The prior references are non-existent.

dcalhoun committed:
Correctly mask Custom URL value when a URL is set via a different option

Selecting Media File or Attachment Page sets a URL value. Originally,
this caused the Custom URL option to display the URL value, rather than
persistently displaying its "Search or type URL" placeholder. The Custom
URL option should only display an actual URL value if it is manually
typed or selected within the link picker.

dcalhoun committed:
Avoid passing current URL to link picker if it is not custom

When the Media File or Attachment Page options are selected, a URL value
is set. Previously, this value was passed to the link picker when the
Custom URL option was subsequently selected. This meant the link picker
would be pre-populated with a (likely unidentifiable) URL that was not
manually entered by the user, which could be confusing.

dcalhoun committed:
Fix stale URL value

Referencing `inputValue` from within LinkSettingsScreen resulted in a
stale value to from `useRoute`. By referencing, the `url` where the
callback is defined, we avoid this issue.

dcalhoun committed:
Fix unnecessary updates to attributes that resulted in bug

Set more appropriate defaults for link settings, the new values match
the sibling state values of the components.  Previously the `undefined`
values for `label` and `rel` differed from the empty string values in
component state. This change helps avoid unnecessary calls to
`setAttributes` that occurred when the BottomSheet was closed.

In addition to being unnecessary, these additional `setAttributes` calls
caused the Image edit component to erroneously change the
`linkDestination` even though the URL had not changed.

dcalhoun committed:
Fix attachment page URL

dcalhoun committed:
Fix incorrect link destination set when toggling new tab or setting rel

When "Open in new tab" or "Link Rel" is modified, the URL is again
passed to be set within `setMappedAttributes`. This caused the
`linkDestination` to improperly be modified. This change now avoid
settings `linkDestination` if the URL is not changing.

dcalhoun committed:
Conditionally display Attachment Page option if URL exists

External images will not have an attachment page URL, so this option
should not be displayed when the URL is undefined.

dcalhoun committed:
Fix alignment for left-aligned selected icon

Toggling the icon opacity rather than the presence of the icon itself
produces the desired alignment for both selected an unselected options.

dcalhoun committed:
Rename style selector to better describe intent

dcalhoun committed:
Remove irrelevant comment regarding setAttributes

After reviewing the logic within `LinkSettings`, it largely manages
invoking `setAttributes` in different scenarios. Therefore, using
`setAttributes` directly makes sense for this component.

dcalhoun committed:
Rename ImageOptionsScreen to ImageLinkDestinationsScreen

Attempt to better describe component intent.

dcalhoun committed:
Add LinkDestination abstraction

Avoid repetition to reduce likelihood of bugs.

dcalhoun commented:
@iamthomasbishop I have created installable builds for [WPiOS](https://github.com/wordpress-mobile/WordPress-iOS/pull/17174#issuecomment-921140720) and [WPAndroid](https://github.com/wordpress-mobile/WordPress-Android/pull/15336#issuecomment-921155161). Would you be willing to review this for design and flow? Thanks! 🙇🏻

iamthomasbishop commented:
@dcalhoun Thanks for spinning up test builds! I had a chance to test and I'm only seeing a few things, only 1 is directly related to this work (all of them apply to both iOS and Android.

The only thing I noticed that is directly related: I'm wondering if we should "hold" the custom URL when the user switches away from using Custom URL to another type and back. Currently doing so clears the previously-used URL, so you have to re-enter it, which I could see being frustrating. 

Another way to solve this would be to list "recently used" URLs when the user taps into the Custom URL field, but I imagine that's not as straight forward of a thing to build.

And _not_ directly related:

- Some of the icon colors seem to be a bit off:
    - The 'checkmark' icon should use our primary blue (`blue-50` in light mode, `blue-30` in dark mode) ([example](https://cloudup.com/czJe-8lWjk2)). Note: this also applies to the checkmark icon on the custom URL sheet
    - The 'back' and 'X' icons seem a bit off, esp in dark mode. IIRC, we should be using `light-secondary` and `dark-secondary` from the [base variables](https://github.com/WordPress/gutenberg/blob/trunk/packages/base-styles/_colors.native.scss)
- **EDIT: Seems this _might_ get resolved by [this PR](https://github.com/WordPress/gutenberg/pull/34425)**. I noticed our bottomSheet still suffers a weird transition when navigating from a child-sheet to its parent. For example, if you are on the Link to sub-sheet (where we list 'none', 'attachment page', etc) and you tap one of the options, the transition upward to its parent is a bit clunky, with the height of the sheet transitioning immediately and its contents transitioning as expected
- If you swipe-to-dismiss the full-height custom URL bottomSheet and then open block settings again, the sheet displays at full-height instead of its standard max-50% height

dcalhoun commented:
> The only thing I noticed that is directly related: I'm wondering if we should "hold" the custom URL when the user switches away from using Custom URL to another type and back. Currently doing so clears the previously-used URL, so you have to re-enter it, which I could see being frustrating.
> 
> Another way to solve this would be to list "recently used" URLs when the user taps into the Custom URL field, but I imagine that's not as straight forward of a thing to build.

@iamthomasbishop thank you for the review. I think some form of caching the most recently used custom URL(s) is likely a great feature. However, given this was a project I attempted (and failed) to finish within a week, I would suggest we postpone that improvement a follow-up iteration. I do not believe the URL caching feature would be insignificant. It is already likely that I will struggle to ship this soon, and adding the caching feature would likely delay shipping further. WDYT? 

> * Some of the icon colors seem to be a bit off:
>   
>   * The 'checkmark' icon should use our primary blue (`blue-50` in light mode, `blue-30` in dark mode) ([example](https://cloudup.com/czJe-8lWjk2)). Note: this also applies to the checkmark icon on the custom URL sheet
>   * The 'back' and 'X' icons seem a bit off, esp in dark mode. IIRC, we should be using `light-secondary` and `dark-secondary` from the [base variables](https://github.com/WordPress/gutenberg/blob/trunk/packages/base-styles/_colors.native.scss)

I will investigate these color issues. If they are quick fixes, then I will include them in this work. 

> * **EDIT: Seems this _might_ get resolved by [this PR](https://github.com/WordPress/gutenberg/pull/34425)**. I noticed our bottomSheet still suffers a weird transition when navigating from a child-sheet to its parent. For example, if you are on the Link to sub-sheet (where we list 'none', 'attachment page', etc) and you tap one of the options, the transition upward to its parent is a bit clunky, with the height of the sheet transitioning immediately and its contents transitioning as expected

Agreed. The navigation transitions are a bit broken currently. I do hope https://github.com/WordPress/gutenberg/pull/34425 resolves some of them, but it is to be seen if it resolves all of them. Also, I still have work to do before I can merge https://github.com/WordPress/gutenberg/pull/34425. Regardless, I think we should address them outside of this PR. 

> * If you swipe-to-dismiss the full-height custom URL bottomSheet and then open block settings again, the sheet displays at full-height instead of its standard max-50% height

Yes, I have noticed this issue as well. I captured it in https://github.com/WordPress/gutenberg/issues/28173. I think we should address it outside of this PR.

dcalhoun committed:
Fix stale link picker value and React Navigation warning

Passing `setAttributes` resulted in a stale value within the link picker
that prohibited setting the same Custom URL twice if attempted without
closing the bottom sheet entirely. E.g. None => Custom URL => None =>
Custom URL.

Additionally, React Navigation warns about passing non-serializable
values as route params, e.g. functions.

These changes addresses both issues.

iamthomasbishop commented:
> However, given this was a project I attempted (and failed) to finish within a week, I would suggest we postpone that improvement a follow-up iteration

Totally fine!

> I will investigate these color issues. If they are quick fixes, then I will include them in this work.

> Agreed. The navigation transitions are a bit broken currently. I do hope #34425 resolves some of them, but it is to be seen if it resolves all of them. Also, I still have work to do before I can merge #34425. Regardless, I think we should address them outside of this PR.

> Yes, I have noticed this issue as well. I captured it in #28173. I think we should address it outside of this PR.

 All of the above sounds good! If we can resolve the color issues, even better!

dcalhoun committed:
Rename image link destination screen for consistency

dcalhoun committed:
Remove Custom URL placeholder

The design of the link destination picker has less width available,
which caused layout issues for the placeholder, particularly on Android
where spacing is larger.

dcalhoun committed:
Reduce spacing around left-side cell icons on Android

This diverges from Material Design guides, but provides the additional
room we require for the new image link destination picker. The Custom
URL option needs to display both an selected icon (checkmark) on the
left side as well as a value and right chevron on the right side.

dcalhoun committed:
Revert "Reduce spacing around left-side cell icons on Android"

This reverts commit 2ef5f650a81f6ceb7238eda79d702d30fc8ce4bd.

dcalhoun committed:
Fix select icon color

dcalhoun committed:
Fix selected icon styles for dark mode

Utilize colors as directed by designer.

dcalhoun committed:
Update selected icon colors

Change to colors requested by designer.

dcalhoun committed:
Merge branch 'trunk' of github.com:WordPress/gutenberg into feature/link-image-to-media

dcalhoun committed:
Update change log

dcalhoun commented on packages/components/src/mobile/link-settings/index.native.js:
```diff
@@ -86,9 +86,9 @@ function LinkSettings( {
 	url,
-	label,
+	label = '',
 	linkTarget,
-	rel,
+	rel = '',
```
These default values were changed to avoid unnecessary attribute updates when the prop transitioned from `undefined` to `''`. The latter is the initial state value used in this component.

The additional, unnecessary attribute updates made it difficult to correctly set the `linkDestination`, which is the focus of this PR. Specifically, having more than one update meant the second update did not have the context (previous `url`) required to set the correct `linkDestination`.

dcalhoun commented on packages/components/src/mobile/link-settings/link-settings-screen.native.js:
```diff
@@ -20,17 +20,21 @@ const LinkSettingsScreen = ( props ) => {
 					props.hasPicker ? onLinkCellPressed : undefined
 				}
 				urlValue={ inputValue }
-				{ ...props }
```
Prop spread relocated to prevent overriding the `onLinkCellPressed` value set here.

mkevins replied:
My understanding is that the `onLinkCellPressed` callback can still be overriden when passed as a prop, but via invocation within the local `onLinkCellPressed` defined above. Is that correct?

Not suggesting a change here, but curious about the scenario where a consumer might clobber this. Did you run into something like that?

dcalhoun replied:
> My understanding is that the `onLinkCellPressed` callback can still be overriden when passed as a prop, but via invocation within the local `onLinkCellPressed` defined above. Is that correct?

Yes, that is correct. 

> Not suggesting a change here, but curious about the scenario where a consumer might clobber this. Did you run into something like that?

If I understand your question correctly, a consumer will always be able to unset (clobber?) `onLinkCellPressed`.

```jsx
<LinkSettingsScreen onLinkCellPressed={false} />

// ...would result in the follow logic in this component...
// onLinkCellPressed={ props.hasPicker ? false : undefined }
```

A consumer will _not_ be able to enforce an `onLinkCellPressed` callback to invoke without setting `hasPicker` to a truthy value. 

```jsx
<>
  <LinkSettingsScreen onLinkCellPressed={() => console.log('> Picker!')} hasPicker={true} />
  <LinkSettingsScreen onLinkCellPressed={() => console.log('> No Picker!')} hasPicker={false} />
</>
// ...would result in...
// > Picker!
```

From searching the project, the usage in this PR is the only context where `onLinkCellPressed` is passed into this component from a parent context. So, I feel the current `hasPicker` logic is safe. WDYT?

mkevins replied:
Thanks for the explanation David! (btw, was using [clobber](https://en.wikipedia.org/wiki/Clobbering) a bit broadly here :smile: )

> From searching the project, the usage in this PR is the only context where onLinkCellPressed is passed into this component from a parent context. So, I feel the current hasPicker logic is safe. WDYT?

I agree.

dcalhoun committed:
Simplify unnecessarily complex conditional

`iconStyle` has a default value of `{}`.

dcalhoun committed:
Reference screen name cache rather than string literal

Attempt to avoid typos.

dcalhoun commented:
>   * The 'checkmark' icon should use our primary blue (`blue-50` in light mode, `blue-30` in dark mode) ([example](https://cloudup.com/czJe-8lWjk2)). Note: this also applies to the checkmark icon on the custom URL sheet

@iamthomasbishop following up to share that I fixed the colors for the checkmark icon to match what you provided—both for the left-side and right-right checkmarks used in _Image Link Destinations_ (this PR) and _Font Size Picker_ (preexisting) UI respectively. So, everything should now match. 

>   * The 'back' and 'X' icons seem a bit off, esp in dark mode. IIRC, we should be using `light-secondary` and `dark-secondary` from the [base variables](https://github.com/WordPress/gutenberg/blob/trunk/packages/base-styles/_colors.native.scss)

I did not make any changes to the back button, but it would appear a recent fix was applied in https://github.com/WordPress/gutenberg/pull/34806, which may address the dark mode issue you observed. 

Once I receive and address code review, I will look to merge this work. Thanks again!

mkevins commented on packages/block-library/src/image/edit.native.js:
```diff
@@ -76,6 +79,89 @@ const getUrlForSlug = ( image, sizeSlug ) => {
+		<PanelBody title={ __( 'Link Settings' ) }>
+			<LinkSettingsNavigation
+				isVisible={ isLinkSheetVisible }
+				url={ mappedAttributes.url }
+				rel={ mappedAttributes.rel }
+				label={ mappedAttributes.label }
+				linkTarget={ mappedAttributes.linkTarget }
```
Some history for context about `mappedAttributes`: originally, the interface to link settings was more narrow, and `attributes` were passed as a "bag of properties" to the component. Most usages included an attribute named `url` for the link url (e.g. format library, button block, etc.), but image block uses the attribute `href` for this. `mappedAttributes` and `setMappedAttributes` was a way to adapt the usage for image block to comply with the API.

This was changed [in a subsequent PR](https://github.com/WordPress/gutenberg/commit/db6d666f3165e78a4dcf350c7ff7f6638e8439e0#diff-7126ae03000a022bcefa0d2a5fb03b6aca56ed4047e3a49d2131babf229cf6c1R380-R383), though, for performance reasons, and the API surface area was expanded to include the attributes individually as props. An opportunity was missed her, imo, to remove the `mappedAttributes` abstraction, as it is no longer necessary.

Since we are refactoring this to an extracted component here, this is a good chance to simplify this, imo, and get rid of `mappedAttributes` (i.e. we can just directly destructure the properties we need from `attributes` [here](https://github.com/WordPress/gutenberg/pull/34846/files#diff-7126ae03000a022bcefa0d2a5fb03b6aca56ed4047e3a49d2131babf229cf6c1R90-R91)). Wdyt?

dcalhoun replied:
Thanks for the suggestion. I addressed this in https://github.com/WordPress/gutenberg/pull/34846/commits/2d7712167ccba131e1bba4f2957d473569516c5d.

mkevins commented on packages/block-library/src/image/edit.native.js:
```diff
@@ -385,43 +452,24 @@ export class ImageEdit extends Component {
+		let linkDestination;
+		if ( restAttributes.linkDestination ) {
+			linkDestination = restAttributes.linkDestination;
+		} else if ( ! href && ! restAttributes.linkDestination ) {
```
Since we are always setting local variable `linkDestination` (thus always overwriting its spread counterpart in `restAttributes`), maybe we can destructure it as a function parameter (along with `href`) instead of introducing the local variable. Wdyt?

mkevins replied:
Won't the second part of this (i.e. `! restAttributes.linkDestination`) always be true here, since we are in an else from the condition previous condition?

so would the following be the same:

```suggestion
		} else if ( ! href ) {
```

dcalhoun replied:
That makes sense to me. I addressed this in https://github.com/WordPress/gutenberg/pull/34846/commits/0749ad210148a99668994d48a11268aa3d92eecc, which changed the conditional logic to still require `&& ! restAttributes.linkDestination`. Let me know if you'd prefer to see this managed in a different manner.

mkevins commented on packages/components/src/mobile/link-settings/style.native.scss:
```diff
@@ -2,3 +2,20 @@
+	color: $blue-30;
+}
+
+.unselectedOptionIcon {
+	opacity: 0;
+}
```
Not suggesting a change here, but curious if this is necessary because of a limitation with the `Cell` component?

mkevins replied:
The reason I ask is that I have in the past encountered "surprises" with the `Cell` component being "too clever", altering its layout from perhaps some assumptions of use cases based on props passed. So I'm wondering if you ran into something similar here. :thinking:

dcalhoun replied:
I'm not sure if I would classify it as "too clever," but for a left-aligned `Cell` if an `icon` is not passed (unselected) then it would "collapse" the space where the icon _would be_ if an icon were passed (selected). So, the end result is that selected and unselected `Cell`s have different left-sided alignment. 

<details><summary>Selected v Unselected Option Alignment</summary>

<img alt="selected-option-alignment" src="https://user-images.githubusercontent.com/438664/134405257-9f9e206f-0b8d-4601-9662-6971fd1ebce5.PNG" width="400" />

</details>

I considered expanding `Cell`'s capabilities to manage this itself, but though it might be premature to add that complexity/conditional to `Cell`. This layout of left-sided checkmark option selection, but also with a right-sided disclosure/value is new and only used here. I thought it might make more see to manage this logic in this one place rather than expand `Cell` now.

mkevins replied:
>  it might be premature to add that complexity/conditional to `Cell`

I agree with this. I think `Cell` is already trying to do too much, imo. Fwiw, your picture jogged my memory on what was surprising: the logic for the separators width was what made assumptions about use cases. In the image you shared above, the "collapsed space" version shows the text component shifted to the left, but the separator remains, so they become misaligned.

mkevins commented on packages/components/src/mobile/link-settings/image-link-destinations-screen.native.js:
```diff
@@ -0,0 +1,147 @@
+			</BottomSheet.NavBar>
+			<PanelBody>
+				<LinkDestination
+					currentOption={ linkDestination }
```
NP: Is the `currentOption` only needed by the `LinkDestination` component to conditionally render the check mark icon? I wonder if the intent would be more "readable" with an `isSelected` boolean prop.

Something like:

`isSelected={ linkDestination === LINK_DESTINATION_NONE }`

Wdyt?

dcalhoun replied:
Yes, that is correct. `currentOption` is solely used for the conditional logic for a selected status. I agree refactoring to `isSelected` is an improvement. Addressed this in https://github.com/WordPress/gutenberg/pull/34846/commits/fa8cea959f4a84dd8000993742652f1fca2849a9.

mkevins commented on packages/components/src/mobile/link-settings/link-settings-screen.native.js:
```diff
@@ -20,17 +20,21 @@ const LinkSettingsScreen = ( props ) => {
 	const onLinkCellPressed = () => {
-		navigation.navigate( 'linkPicker', { inputValue } );
+		if ( props.onLinkCellPressed ) {
+			props.onLinkCellPressed( { navigation } );
```
Also, not suggesting a change here, just curious about the scenario where the url was stale. I saw this mentioned in this commit: https://github.com/WordPress/gutenberg/commit/9a806c355d9fe9c9dc346bdb1707cc8914a38f61, and I'm wondering how it came to be stale, since `inputValue` is in the dependencies array of the `useMemo` :thinking: 

Reason I'm curious is that I have noticed some regressions in the link picker (not in your PR) and haven't tracked the root causes down yet. What you encountered might shed light on some possible causes.

dcalhoun replied:
Unfortunately, I do not have a good, clean answer for you regarding the stale value. It was quite puzzling. With https://github.com/WordPress/gutenberg/commit/9a806c355d9fe9c9dc346bdb1707cc8914a38f61 I attempted to work around the issue by using a different value. 

However, the stale `inputValue` issue became apparent again in a different context later. That discovery—as well as a need to avoid [a React Navigation warning](https://reactnavigation.org/docs/troubleshooting/#i-get-the-warning-non-serializable-values-were-found-in-the-navigation-state)—led me to refactor the overall approach in https://github.com/WordPress/gutenberg/commit/d850dd35dc7931b09a835fb4cd316b580ff05d09. Not only did I relocate attribute updating to avoid passing a function in route params, I also [explicitly reset the `inputValue`](https://github.com/WordPress/gutenberg/commit/d850dd35dc7931b09a835fb4cd316b580ff05d09#diff-1925dc0bdda08deb13d65bbf9a16ec62d9905146b6bebf59734048261d999403R85-R86) when a new link destination was selected. 

So, the staleness could be related to my original implementation ignoring React Navigation's warning about non-serializable route params, but I my hunch is that it is more likely related to lexical scope and the parent-child relationship. The callback is defined in a parent render, the child is the one invoking the callback with (originally) route params.

mkevins replied:
> Unfortunately, I do not have a good, clean answer for you regarding the stale value.

Actually, your comment is very helpful, so thanks for the extra context! Regarding the function as a route param or the binding for the passed function, it sounds like the combination of these things could lead to inconsistent state. I'm glad you found a solution that works :+1: .

dcalhoun committed:
Remove unnecessary mappedAttributes assignment

Now that we pass each prop individually to the child component, it is
arguably simpler to destructure the props required.

dcalhoun committed:
Reuse linkDestination parameter instead of new local assignment

Reassigning a destructured function parameter, instead of creating a new
local assignment, allows for fewer conditionals and less code.

dcalhoun committed:
Refactor selected status of LinkDestination

Relocate the logic to allow for a single `isSelected` prop to improve
readability.

dcalhoun commented:
@mkevins thank you for the review. 🙇🏻 I have addressed each of your comments. When you have time, will you please review again and let me know if there is anything we'd like to address further?

mkevins commented:
Thanks for addressing all the comments!

> When you have time, will you please review again and let me know if there is anything we'd like to address further?

I've given it another pass, and it is looking great! Also, I tested it out, and it is mostly working according to the steps described, with a few exceptions:

For some reason, at one point I got into a state where the media file option would not persist. The steps are as follows:

1. Set a custom URL
2. Select none
3. Select media file

When I did this, the media file was not set, and instead it remained at none. :thinking: 

<video src="https://user-images.githubusercontent.com/8507675/134455581-4b2da4c7-1e7c-433a-bea8-b5f6c501d767.mp4" width="360"></video>

When I set the media URL to an external image (via the web) I happened to have the attachment page setting already set. This resulted in the attachment page appearing to be set, but it was not present in the options (as expected).

<video src="https://user-images.githubusercontent.com/8507675/134455661-abd1a0e9-b9d2-4352-917d-df8f97d98d91.mp4" width="360"></video>

I'm not sure if this is something we can easily address (and may likely be an issue for web as well). Also, it's a bit "out of the way" to reproduce it, so I wouldn't expect the impact to be large for this issue. Just noting it since I happened to encounter this while testing that flow.

Finally (though this is not introduced in this PR, and I'm only mentioning for completeness), I also encountered the strange "transparent fullscreen modal" again:

<img src="https://user-images.githubusercontent.com/8507675/134455848-4893d600-5707-4b09-a5ad-bd9bfd783f17.png" width="360">

dcalhoun commented:
> For some reason, at one point I got into a state where the media file option would not persist. The steps are as follows:
> 
> 1. Set a custom URL
> 2. Select none
> 3. Select media file
> 
> When I did this, the media file was not set, and instead it remained at none. 🤔

@mkevins I observed an issue like this in the past, but I am been unable to reproduce this now. 😕 May I verify you are using the latest [build for Android](https://github.com/wordpress-mobile/WordPress-Android/pull/15336#issuecomment-921155161): `WordPress-pr-15336-build-115665.apk`? If yes, were there any other steps to reproduce? New image or existing image in a post? Did you close the bottom sheet between switching Link Destinations? Update the post between switching options?

> When I set the media URL to an external image (via the web) I happened to have the attachment page setting already set. This resulted in the attachment page appearing to be set, but it was not present in the options (as expected).
> 
> I'm not sure if this is something we can easily address (and may likely be an issue for web as well). Also, it's a bit "out of the way" to reproduce it, so I wouldn't expect the impact to be large for this issue. Just noting it since I happened to encounter this while testing that flow.

From my testing, it does appear the web suffers from the same "stale" link issue. However, it is not as pronounced as the web UI merely renders the "stale" link address rather than claiming it to be the image's "Attachment Page" specifically. 

<details><summary>Web UI: Image with "detached" Attachment Page URL</summary>
<img width="590" alt="web-image-with-detached-url" src="https://user-images.githubusercontent.com/438664/134568010-7cf8969a-3b89-484d-b623-c9a21375a902.png">
</details>

One fix might be no longer relying solely upon `linkDestination` to dictate the currently selected option. We could check if the set `url` is _actually_ the attachment page URL. It feels a bit "icky" to no longer treat `linkDestination` as truth, but I wonder if that is what the web is doing. WDYT? 

I would agree—with what I believe you implied—that we could ship with this issue as it is likely an edge case. 
 
> Finally (though this is not introduced in this PR, and I'm only mentioning for completeness), I also encountered the strange "transparent fullscreen modal" again:

Yes, hopefully we are able to find a solution to this issue within https://github.com/WordPress/gutenberg/pull/34425.

mkevins commented:
> @mkevins I observed an issue like this in the past, but I am been unable to reproduce this now.  May I verify you are using the latest [build for Android](https://github.com/wordpress-mobile/WordPress-Android/pull/15336#issuecomment-921155161): `WordPress-pr-15336-build-115665.apk`? If yes, were there any other steps to reproduce? New image or existing image in a post? Did you close the bottom sheet between switching Link Destinations? Update the post between switching options?

Actually, I encountered this issue by testing this PR and using WordPress-Android develop (at this commit: https://github.com/wordpress-mobile/WordPress-Android/commit/9a6f1154a1aa4ee6177843a3f0de8106753a2b80). It was with a new image that I created by inserting an image block and choosing an image from the media library. I did not close the bottomsheet between switching link destinations, nor update the post between switching options. I will also try with the generated APK to see if there's different behavior there.

> I would agree—with what I believe you implied—that we could ship with this issue as it is likely an edge case.

Yes, you believe right :smile:  : I don't consider the second issue to be a blocker for merging the PR.

mkevins commented:
I was able to reproduce it on the apk as well (though it seems I had to fiddle quite a bit by changing the options a lot — I even added custom URL twice). I'm not sure why sometimes it happens with fewer steps :thinking: 

<video src="https://user-images.githubusercontent.com/8507675/134631252-336d7b0c-a656-4318-b0b0-f57435f38ea8.mp4"></video>

You may also notice in that video a regression unrelated to this PR in which the search results are cleared a moment after they are displayed, which seems as though it might be related with the keyboard suggestion, since it happens to me when I swipe a word, but not hunt-and-peck :man_shrugging: . I wonder whether these kinds of state issues could be related with the bottomsheet animations / navigation :thinking:

dcalhoun committed:
Remove unused code

There are no references to this method in the source.

dcalhoun committed:
Remove undefined callback

This prop referenced an undefined class method.

dcalhoun committed:
Add Image Link To tests

dcalhoun committed:
Fix incorrect Link To value mask caused by stale route parameters

Originally, ImageLinkDestinationsScreen relied upon (1) differing route
parameter names from LinkPicker and (2) explicitly set `linkDestination`
while LinkPicker did not. Stale route parameters resulted in erroneous
attribute updates when swapping the link destination from Media File >
Custom URL > Media File.

This change addresses the stale route parameters by (1) unifying the
route parameter names between the two components and (2) unifying the
approach for setting `linkDestination` to be computed based upon the
URL. The URL, represented as the `inputValue` route parameter, is now
the sole route parameter.

dcalhoun committed:
Merge branch 'trunk' of github.com:WordPress/gutenberg into feature/link-image-to-media

dcalhoun committed:
Fix erroneous URL display within Custom URL input

The image URL may contain query parameters for aspects like display
dimensions. This caused a comparison based upon the URL to unexpectedly
mismatch. By leveraging `linkDestination` instead, we correctly hide the
URL from the Custom URL input whenever Media File or Attachment Page is
selected.

dcalhoun committed:
Format initialHtml seed consistently

dcalhoun commented:
@mkevins thank you for your patience while I found time to address your feedback. 🙇🏻 This PR and [its sibling](https://github.com/wordpress-mobile/gutenberg-mobile/pull/3971) are now ready for another round of review. Installable builds can be found on the sibling [WPiOS](https://github.com/wordpress-mobile/WordPress-iOS/pull/17174) and [WPAndroid](https://github.com/wordpress-mobile/WordPress-Android/pull/15336) PRs. 

I ended up cherry picking a few commits from another branch to allow me to write tests from this new Image work. So, it may be easiest to step through each of the commits since your last review individually. I tried to leave commit messages explaining each. 

## (FIXED) Selection Sporadically Not Persisting

I believe I have addressed this issue within https://github.com/WordPress/gutenberg/pull/34846/commits/cf4027fcb3c0d41acf060ffd72e119b9391915c9. To recap, here are the steps I used to reproduce the issue:

1. Add an Image block and set its media. 
2. Change _Link To_ from _None_ to _Media File_. 
3. Change _Link To_ from _Media File_ to _Custom URL_.
4. Change _Link To_ from _Custom URL_ to _Media File_. 
5. 💥 The _Link To_ is now erroneously displayed as _None_.  

## (POSTPONED) Incorrect Link To After Swap to External Image

[We agreed](https://github.com/WordPress/gutenberg/pull/34846#issuecomment-926384423) that we do no need to address this issue within this PR, but I will recap in case we decide to re-engage it later. We noticed the web editor seemingly suffers from a stale URL in this context as well, but it is not as noticeable. For future reference [these lines](https://github.com/WordPress/gutenberg/blob/154b34a8f5dff5fec056e0e6941fa3aa141ba823/packages/block-library/src/image/edit.js#L181-L218) appear to be relevant, and were modified from the previous [code that managed](https://github.com/WordPress/gutenberg/blob/2149b4eb5a1a5775ba60ffbad336b98b714c869b/packages/block-library/src/image/edit.js#L155-L165) this scenario. I will create a new issue for this if/when we merge this work. 

1. In the mobile editor, add an Image block and set its media. 
2. Open the same post in the web editor, change the Image block's media to an external image URL. 
3. In the mobile editor, view the same Image block again. 
4. 💥 Notice the _Link To_ claims to be _Attachment Page_, but that is not valid for an external image. The _Attachment Page_ option is not available within the options.

mkevins reviewed (approved):
Thanks for addressing the issues David! I also very much appreciate the attention to detail and the added and improved tests :+1: . The code looks good, and I have confirmed that the "link to not persisting" bug has been resolved. Nice work!
